drools-parser/src/main/java/org/drools/parser/DRLVisitorImpl.java

```diff
@@ -125,80 +118,98 @@ public Object visitRuledef(DRLParser.RuledefContext ctx) {
+            return getSinglePatternDescr(ctx);
         } else if (ctx.lhsPattern().size() > 1) {
-            OrDescr orDescr = new OrDescr();
-            currentConstructStack.peek().addDescr(orDescr);
-            currentConstructStack.push(orDescr);
-            try {
-                Object result = super.visitLhsPatternBind(ctx);
-                List<? extends BaseDescr> descrs = orDescr.getDescrs();
-                for (BaseDescr descr : descrs) {
-                    PatternDescr patternDescr = (PatternDescr) descr;
-                    if (ctx.label() != null) {
-                        patternDescr.setIdentifier(ctx.label().IDENTIFIER().getText());
-                    }
-                }
-                return result;
-            } finally {
-                currentConstructStack.pop();
-            }
+            return getOrDescrWithMultiplePatternDescr(ctx);
         } else {
             throw new IllegalStateException("ctx.lhsPattern().size() == 0 : " + ctx.getText());
         }
     }
 
+    private PatternDescr getSinglePatternDescr(DRLParser.LhsPatternBindContext ctx) {
+        Optional<BaseDescr> optPatternDescr = visitFirstChild(ctx);
+        PatternDescr patternDescr = optPatternDescr.filter(PatternDescr.class::isInstance)
+                .map(PatternDescr.class::cast)
+                .orElseThrow(() -> new IllegalStateException("lhsPatternBind must have at least one lhsPattern : " + ctx.getText()));
+        if (ctx.label() != null) {
+            patternDescr.setIdentifier(ctx.label().IDENTIFIER().getText());
+        }
+        return patternDescr;
+    }
+
+    private OrDescr getOrDescrWithMultiplePatternDescr(DRLParser.LhsPatternBindContext ctx) {
+        OrDescr orDescr = new OrDescr();
+        List<BaseDescr> patternList = visitChildrenWithDescrAggregator(ctx, new ArrayList<>());
+        patternList.stream()
+                .filter(PatternDescr.class::isInstance)
+                .map(PatternDescr.class::cast)
+                .forEach(patternDescr -> {
+                    if (ctx.label() != null) {
+                        patternDescr.setIdentifier(ctx.label().IDENTIFIER().getText());
+                    }
+                    orDescr.addDescr(patternDescr);
+                });
+
+        return orDescr;
+    }
+
     @Override
-    public Object visitLhsPattern(DRLParser.LhsPatternContext ctx) {
-        currentPattern = new PatternDescr(ctx.objectType.getText());
+    public PatternDescr visitLhsPattern(DRLParser.LhsPatternContext ctx) {
+        PatternDescr patternDescr = new PatternDescr(ctx.objectType.getText());
+        currentPattern = patternDescr;
         if (ctx.patternSource() != null) {
             String expression = ctx.patternSource().getText();
             FromDescr from = new FromDescr();
             from.setDataSource(new MVELExprDescr(expression));
             from.setResource(currentPattern.getResource());
             currentPattern.setSource(from);
```

Comment:
Hi @tkobayas !
I'm not sure to follow the logic of `currentPattern.setSource(from);` (@187) and `currentPattern = null;` (@190) and, in general, of this method: at line @101 `currentPattern` points to same object as `patternDescr`; in the `if` statement the status of that object is changed (through `currentPattern`), and then `currentPattern`reference is nullified: I have the impression that that `currentPattern` is not needed here and, in general, being it a "status" of the class, make the code slightly harder to understand/maintain then needed: wdyt ?